### PR TITLE
Improved docs for infer_datetime_format

### LINKED ISSUE #12152: better docs for read_csv() argument infer_datetime_format
This gives pretty decent explanation of the infer_datetime_format parameter:

http://pandas.pydata.org/pandas-docs/version/0.17.1/whatsnew.html#id55

The api docs are fairly opaque.

Improve api docs to make it clearer:

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html

### PR DESCRIPTION
**Pull request by robintw**
Fixes #12152

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -132,8 +132,10 @@ class ParserWarning(Warning):
 
     Note: A fast-path exists for iso8601-formatted dates.
 infer_datetime_format : boolean, default False
-    If True and parse_dates is enabled for a column, attempt to infer
-    the datetime format to speed up the processing
+    If parse_dates is enabled and this flag is set, pandas will attempt to
```
Comment (jreback) [inline pandas/io/parsers.py]: add ``if True``

**Comment by jreback** [comment]
pls add to doc-string of ``pd.to_datetime`` as well.

**Comment by robintw** [comment]
Thanks - I think I've sorted all of those now.

**Comment by jreback** [comment]
maybe show an example where this helps (in Examples) mainly for pd.to_datetime

File: pandas/tseries/tools.py [hunk 2]
```diff
@@ -273,6 +275,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     99   2000-04-09
     Length: 100, dtype: datetime64[ns]
 
+    Infer the format from the first entry
+
+    >>> pd.to_datetime(df.month + '/' +  df.day + '/' + df.year,
+                       infer_datetime_format=True)
```
Comment (jreback) [inline pandas/tseries/tools.py]: this is not valid code (as ``df.month`` are integers), do something like: 

``s = df.month.astype(str) + '/' +  df.day.astype(str) + '/' + df.year.astype(str)``

then infer on that

File: pandas/tseries/tools.py [hunk 3]
```diff
@@ -273,6 +275,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
...
+    0    2000-01-01
+    1    2000-01-02
+    ...
+    98   2000-04-08
+    99   2000-04-09
+
+    This gives the same results as omitting the `infer_datetime_format=True`,
```
Comment (jreback) [inline pandas/tseries/tools.py]: this is confusing, pls reword

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -132,8 +132,10 @@ class ParserWarning(Warning):
 
     Note: A fast-path exists for iso8601-formatted dates.
 infer_datetime_format : boolean, default False
-    If True and parse_dates is enabled for a column, attempt to infer
-    the datetime format to speed up the processing
+    If True and parse_dates is enabled, pandas will attempt to infer the format
+    of the datetime strings in the columns, and if it can be inferred, switch
+    to a faster method of parsing them. In some cases this can increase the
+    parsing speed by ~5-10x.
```
Comment (jreback) [inline pandas/io/parsers.py]: put in a link to ``pd.to_datetime`` here

File: pandas/tseries/tools.py (hunk 2 above)
Comment (robintw) [inline pandas/tseries/tools.py]: I assumed that the `df = df.astype(str)` from the example above would still have applied.

Should each example be entirely separate?

Comment (jreback) [inline pandas/tseries/tools.py]: ahh I see that. ok no, that is fine then (maybe rename ``df`` -> ``dfs`` or something

**Comment by jreback** [comment]
thanks!

File: pandas/io/parsers.py (hunk 4 above)
Comment (jreback) [inline pandas/io/parsers.py]: hah, I forgot my own comment. If you want to add this in another PR great.

**Pull request closed without merging**